lib/db/schema/patch-022-023.sql

```diff
@@ -0,0 +1,263 @@
+    uaOS,
+    uaOSVersion,
+    uaDeviceType,
+    createdAt
+  );
+
+  IF tokenVerified = FALSE THEN
+    INSERT INTO unverifiedTokens(
+      tokenVerificationId,
+      tokenId
+    )
+    VALUES(
+      tokenVerificationId,
+      tokenId
+    );
+  END IF;
+
+  COMMIT;
+END;
+
+-- Update createKeyFetchToken stored procedure to
+-- insert into unverifiedTokens.
+CREATE PROCEDURE `createKeyFetchToken_2` (
+  IN tokenId BINARY(32),
+  IN authKey BINARY(32),
+  IN uid BINARY(16),
+  IN keyBundle BINARY(96),
+  IN createdAt BIGINT UNSIGNED,
+  IN tokenVerified BOOLEAN,
+  IN tokenVerificationId BINARY(32)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  INSERT INTO keyFetchTokens(
+    tokenId,
+    authKey,
+    uid,
+    keyBundle,
+    createdAt
+  )
+  VALUES(
+    tokenId,
+    authKey,
+    uid,
+    keyBundle,
+    createdAt
+  );
+
+  IF tokenVerified = FALSE THEN
+    INSERT INTO unverifiedTokens(
+      tokenVerificationId,
+      tokenId
+    )
+    VALUES(
+      tokenVerificationId,
+      tokenId
+    );
+  END IF;
+
+  COMMIT;
+END;
+
+-- Update deleteSessionToken stored procedure to
+-- conditionally delete from unverifiedTokens.
+CREATE PROCEDURE `deleteSessionToken_2` (
+  IN tokenIdArg BINARY(32)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM sessionTokens WHERE tokenId = tokenIdArg;
+  DELETE FROM unverifiedTokens WHERE tokenId = tokenIdArg;
+
+  COMMIT;
+END;
+
+-- Update deleteKeyFetchToken stored procedure to
+-- conditionally delete from unverifiedTokens.
+CREATE PROCEDURE `deleteKeyFetchToken_2` (
+  IN tokenIdArg BINARY(32)
+)
+BEGIN
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  DELETE FROM keyFetchTokens WHERE tokenId = tokenIdArg;
+  DELETE FROM unverifiedTokens WHERE tokenId = tokenIdArg;
+
+  COMMIT;
+END;
+
+-- Add stored procedure for fetching sessionToken
+-- with its verified state.
+CREATE PROCEDURE `sessionTokenVerified_1` (
+  IN tokenIdArg BINARY(32)
+)
+BEGIN
+  DECLARE tokenVerified BOOLEAN;
+
+  DECLARE EXIT HANDLER FOR SQLEXCEPTION
+  BEGIN
+    ROLLBACK;
+    RESIGNAL;
+  END;
+
+  START TRANSACTION;
+
+  SET tokenVerified = (
+    CASE WHEN EXISTS(
+      SELECT * FROM unverifiedTokens
+      WHERE unverifiedTokens.tokenId = tokenIdArg
+    )
+      THEN FALSE
+      ELSE TRUE
+    END
+  );
+
+  SELECT
+    t.tokenData,
+    t.uid,
+    t.createdAt,
+    t.uaBrowser,
+    t.uaBrowserVersion,
+    t.uaOS,
+    t.uaOSVersion,
+    t.uaDeviceType,
+    t.lastAccessTime,
+    a.emailVerified,
+    a.email,
+    a.emailCode,
+    a.verifierSetAt,
+    a.locale,
+    a.createdAt AS accountCreatedAt,
+    tokenVerified AS tokenVerified
+  FROM sessionTokens AS t
+  INNER JOIN accounts AS a
+  ON t.uid = a.uid
+  WHERE t.tokenId = tokenIdArg;
+
+  COMMIT;
+END;
```

Comment:
ok. I found the IRC conversation, so I understand the logic, but I still don't see why we need more than `unverifiedTokens` (tokenId, tokenVerificationId). Before I get into it here's the IRC thread:

```
4:17 AM <rfkelly> it's not obvious to me how we represent the distinction between "a token that was verified via email challenge" and "a token for which we never sent an email challenge in the first place"
4:18 AM <rfkelly> i.e. if I get a sessionToken claiming to log in to an oauth relier, will sessionTokenVerified_1 tell me that it's verified, because it doesn't exist in the table?
4:18 AM <pb> rfkelly: yep
4:18 AM <pb> rfkelly: is my mental model broken?
4:19 AM <pb> rfkelly: why do we need to distinguish between those two cases?
4:20 AM <rfkelly> maybe we don't, as long as the keyFetchTokens are also verified
4:20 AM <rfkelly> but if the requirement is something like "you can only log into sync with a verified session token" then it seem a bit odd
4:21 AM <rfkelly> that I can create sessionTokens that this procedure will claim are verified, but actually aren't verified
4:22 AM <rfkelly> perhaps what I'm struggling to express there is, could I generate a no-verification-email sessionToken by claiming I'm logging into an oauth provider, and then use that sessionToken to authenticate to sync?
4:22 AM <pb> rfkelly: ah, i see your point now. at the very least it is misleading.
4:22 AM <rfkelly> also maybe I should go read the latest version of the feature doc where this may be covered :-)
4:23 AM <pb> rfkelly: arrgh, no it isn't covered, you're completely right and i am a fool :)
4:24 AM <rfkelly> as I said, maybe that's ok as long as the keyFetchToken is properly verified, as you need both to actually log in to sync
4:24 AM <rfkelly> but it seems fragile
4:25 AM <rfkelly> (actually this whole "table of unverified tokens" thing seems fragile to me as it's easy for it to "fail open" like in the above scenario - although I understand the appeal from an implementation POV)
4:25 AM ⇐ Standard8 quit (Standard8@moz-1hu.vrc.166.195.IP) Ping timeout: 121 seconds
4:26 AM <rfkelly> it's hard when we also have the requirement that existing tokens should be "verified" by default :-/
4:27 AM <pb> rfkelly: yep, that is the constraint that led me down this path
4:28 AM <rfkelly> *nod*
4:28 AM <pb> rfkelly: so this doesn't address your concern, but it is related: in the first version of the PR, keyFetchTokens were always created in the unverified state. then i changed it to let the auth server decide as it does for session tokens. should i go back to how it was?
4:28 AM <rfkelly> under what circumstances would you create an already-verified sessionToken?
4:29 AM <pb> rfkelly: when request.query.keys !== 'TRUE'
4:29 AM <pb> s/TRUE/true/
4:30 AM <rfkelly> so, that's more-or-less the same thing as I was talking about above...could I create one of those, and then turn around and actually use it to log in to sync?
4:30 AM <rfkelly> (obviously I'd also have to get the keys from somewhere, but still...)
4:30 AM <pb> rfkelly: yep, and this is where my lack of knowledge about the wider system leaves me floundering a bit
4:31 AM <rfkelly> yeah we have a lot, lot, lot of tokens doing different things in this system
4:31 AM <rfkelly> non-obvious things
4:31 AM <pb> rfkelly: okay, different question
4:32 AM <pb> rfkelly: if we create every new session token in the unverified state, can we safely auto-verify tokens for oauth reliers somehow so that those users don't have to confirm their email?
4:33 AM <pb> rfkelly: that seemed like it was impossible to me hence the only creating unverified when i do
4:33 AM <rfkelly> hmm...so let me respond to your question with another question :-)
4:34 AM <rfkelly> why do you need a token in 'verified' state in order to log in to an oauth relier?
4:34 AM <rfkelly> I assume the answer to be "so you can call /certificate/sign"
4:34 AM <rfkelly> but bad things happen when I assume
4:35 AM <pb> rfkelly: ummm, it is an excellent question. thinking...
4:36 AM <pb> rfkelly: so i think also, when we modeled this originally, we were thinking about when other endpoints are protected against unverified tokens too. e.g. /account/destroy we didn't want to stop people doing that
4:36 AM <rfkelly> fwiw I think this is what I was trying to get at in my comments here as well: https://github.com/mozilla/fxa/pull/143#discussion_r59347940
4:38 AM <rfkelly> my suggestion is broadly:
4:38 AM <pb> rfkelly: ah, okay, there are pennies dropping in my head
4:38 AM <rfkelly> 1) make /account/keys require a verified token, always
4:39 AM <rfkelly> 2) allow /certificate/sign to be called with any type of sessionToken, but record the verified status of the token in the generated certificate
4:39 AM <rfkelly> and separately to all that
4:39 AM <rfkelly> 3) make sync reject certificate that didn't come from a verified token
4:40 AM <pb> rfkelly: okay, that all makes sense, 2) was the bit i hadn't properly considered when i read your github comment first time
4:41 AM <rfkelly> my follow-up waffling about OIDC probably didn't help either :-)
4:41 AM <pb> rfkelly: re 3), does that violate our no-need-for-client-changes constraint?
4:41 AM <rfkelly> no, we can enforce that on the server-side in sync
4:41 AM <pb> rfkelly: cool beans
4:42 AM <pb> rfkelly: thanks, lots of things that weren't clear, now are
4:42 AM <rfkelly> in theory, either (1) or (2) is sufficient to stop an attacker from getting into your sync data without an email confirmation
```

The 3 states @rfk mentioned earlier don't apply here. As far as I can tell a token still only needs to be either unverified or "not unverified" (legacy FTW). My question: *"Wouldn't we simply not create a unverifiedTokens record in cases we don't care to verify?"* is moot. We **only** create unverified tokens, but we give them a bit more power, i.e. allow `/certificate/sign` and note the state in the response. That gives the relier enough info to decide how to act.